Fix issue with launch grid calculation

nandesuka opened the pull request:
Summary: per title

Test Plan:
CI

Rollback Plan:

Differential Revision: D79186985




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

jingsh reviewed (commented):
can you add tests to cover this change?

nandesuka committed:
Fix issue with launch grid calculation (#159388)

Summary:

per title

Test Plan:
CI

Rollback Plan:

Differential Revision: D79186985

aorenste commented on torch/_inductor/runtime/triton_heuristics.py:
```diff
@@ -2961,8 +2961,7 @@ def ceildiv(
-            return f"-(({numel}) // -({block}))"
-        # trick above doesn't work in C++ due to rounding differences
+            return f"(({numel} + {block} - 1) // {block})"
         return f"(({numel} + ({block} - 1)) / ({block}))"
```
Wouldn't that change the desired behavior? This is the integer version of `ceil(numel / block)`...

blaine-rister reviewed (changes requested):
Glad we are fixing issues with dynamic shapes. We actually tried a similar approach before with `FloorDiv`, but it caused a performance regression on a few models, so we switched to the current approach of pattern matching FloorDiv inside the FX converter. It should be possible to do a similar pattern match for `CeilDiv`. There's some context in https://github.com/pytorch/pytorch/pull/153230. 

Taking a step back, I'm also wondering if the current python codegen is actually wrong, or if we're just processing the grid expression incorrectly somewhere else in the stack. We already have some similar dynamic shape tests on GPU, so it's possible something weird is going on with MTIA. I could be wrong, but just wanted to double check. See here for example: https://github.com/pytorch/pytorch/blob/main/test/inductor/test_fxir_backend.py#L359

cc @jansel

blaine-rister commented on test/inductor/test_fxir_backend.py:
```diff
@@ -411,6 +411,28 @@ def test_dynamic_shapes_precomputed_size(self):
+        self.assertIn("xnumel", triton_node.kwargs["kwargs"])
+        grid = triton_node.kwargs["grid"][0]
+        xnumel = triton_node.kwargs["kwargs"]["xnumel"]
+        self.assertEqual(grid[0].node.expr.subs(xnumel, 1), 1)
```
It seems like this depends on the block size heuristics. Is there a way to make the test generalize to other block sizes? Maybe we should check that the grid is equal to `ceil(xnumel / XBLOCK)`?

nandesuka commented:
Abandoning for a different solution
